### PR TITLE
Sema: Don't crash when type-checking a KeyPathExpr that's already type-checked.

### PR DESCRIPTION
**Pull request by jckarter**
This can come up when re-checking a larger expression for diagnostics. Fixes SR-4965.
rdar://problem/32328705

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please smoke test

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -3988,6 +3988,20 @@ namespace {
...
+      // resolved, and we can let the expression be. This might happen when
+      // re-checking a failed system for diagnostics.
+      if (!E->getComponents().empty()
+          && E->getComponents().front().isResolved()) {
+        assert([&]{
+          for (auto &c : E->getComponents())
+            if (!c.isResolved())
+              return false;
+          return true;
+        }());
```
Comment (jrose-apple) [inline lib/Sema/CSApply.cpp]: Nitpick: `llvm::all_of` would work very nicely here.

Comment (jckarter) [inline lib/Sema/CSApply.cpp]: I did try that, but the lambda noise felt like it outweighed any benefit over Just Writing A For Loop.

Comment (jrose-apple) [inline lib/Sema/CSApply.cpp]: Hm. To me the immediately-called lambda is already weird enough to counter that, but okay.

```c++
assert(llvm::all_of(E->getCompanents(),
                    [](const KeyPathExpr::Component &c) {
  return c.isResolved();
}));
```

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Pull request merged**